src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,258 @@
+/// into the memory of other constants or statics
+struct IsStaticOrFn;
+
+impl<'rt, 'a, 'mir, 'tcx> InternVisitor<'rt, 'a, 'mir, 'tcx> {
+    fn intern(
+        &mut self,
+        ptr: Pointer,
+        mutability: Mutability,
+    ) -> EvalResult<'tcx, Option<IsStaticOrFn>> {
+        trace!(
+            "InternVisitor::intern {:?} with {:?}",
+            ptr, mutability,
+        );
+        // remove allocation
+        let tcx = self.ecx.tcx;
+        let memory = self.ecx.memory_mut();
+        let (kind, mut alloc) = match memory.alloc_map.remove(&ptr.alloc_id) {
+            Some(entry) => entry,
+            None => {
+                // if the pointer is dangling (neither in local nor global memory), we leave it
+                // to validation to error. The `delay_span_bug` ensures that we don't forget such
+                // a check in validation.
+                if tcx.alloc_map.lock().get(ptr.alloc_id).is_none() {
+                    tcx.sess.delay_span_bug(self.ecx.tcx.span, "tried to intern dangling pointer");
+                }
+                // treat dangling pointers like other statics
+                // just to stop trying to recurse into them
+                return Ok(Some(IsStaticOrFn));
+            },
+        };
+        // This match is just a canary for future changes to `MemoryKind`, which most likely need
+        // changes in this function.
+        match kind {
+            MemoryKind::Stack | MemoryKind::Vtable => {},
+        }
+        // Ensure llvm knows to only put this into immutable memory if the value is immutable either
+        // by being behind a reference or by being part of a static or const without interior
+        // mutability
+        alloc.mutability = mutability;
+        // link the alloc id to the actual allocation
+        let alloc = tcx.intern_const_alloc(alloc);
+        self.leftover_relocations.extend(alloc.relocations.iter().map(|&(_, ((), reloc))| reloc));
+        tcx.alloc_map.lock().set_alloc_id_memory(ptr.alloc_id, alloc);
+        Ok(None)
+    }
+}
+
+impl<'rt, 'a, 'mir, 'tcx>
+    ValueVisitor<'a, 'mir, 'tcx, CompileTimeInterpreter<'a, 'mir, 'tcx>>
+for
+    InternVisitor<'rt, 'a, 'mir, 'tcx>
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+    fn visit_aggregate(
+        &mut self,
+        mplace: MPlaceTy<'tcx>,
+        fields: impl Iterator<Item=EvalResult<'tcx, Self::V>>,
+    ) -> EvalResult<'tcx> {
+        if let Some(def) = mplace.layout.ty.ty_adt_def() {
+            if Some(def.did) == self.ecx.tcx.lang_items().unsafe_cell_type() {
+                // we are crossing over an `UnsafeCell`, const qualif ensures that this can only
+                // exist in the base constant, not behind references. Double checking here to make
+                // sure everything is alright
+                assert_eq!(self.mutability, Mutability::Mutable);
```

Comment:
I was just confused for a bit how we manage to forbid `UnsafeCell` behind references in a `static` without forbidding top-level `UnsafeCell`. That is the reason why this check is in `visit_aggegrate` and not in `visit_primitive`, right? Long-term that's wrong though, for `const` we want to check for `UnsafeCell` in `visit_primitive` (or wherever it would also check the top-level), and for `static` we could pefectly well allow `UnsafeCell` even behind references.

Also please add a message here saying what's wrong, so that if someone runs into the ICE we have a bit of a clue where it comes from even if there is no debug info.

Reply:
Hm no that's not right, if the type is `Cell<T>` we will still hit this code path.

This is just a sanity check that whoever determined the mode of the allocation we are currently walking did their job right by checking "frozen", right?

Reply:
>  That is the reason why this check is in visit_aggegrate and not in visit_primitive, right? Long-term that's wrong though, for const we want to check for UnsafeCell in visit_primitive (or wherever it would also check the top-level), and for static we could pefectly well allow UnsafeCell even behind references.

This works because the base constant checks whether the outermost type is frozen and sets the `mutability` field accordingly.

Reply:
Yeah I realized later that I was wrong, see my second message. :)  I think the term "base constant" you used here lead me totally astray.

Reply:
So could you please change the comment to say something like

> We are crossing over an `UnsafeCell`; it is the responsibility of whoever set up this visitor to make sure that our mutability is set appropriately. Double checking here to make sure everything is alright.

The "const qualif forbids nested `UnsafeCell`" part is necessary elsewhere (namely in the recursive case to justify that we don't look at "frozen").